### PR TITLE
Warming views, legacy support

### PR DESCRIPTION
**Pull request by SCdF**
Stage 2 of horticulturalist improvements, for your consideration.

 - Adds support for detecting legacy deploy docs and converting them to the new style (before deploying as normal). This means you can deploy from master -> 2.14.0 and back again, and horti just works. I'm probably going to add a `version` property to the `build_info` (and maybe `deploy_info`) structures so we can validate that you've passed a version, and we can switch on that in the future if we have to. Can even bind the version property to horti versions (ie if there is a `"build_info": {"version": 2 ... }` you know that means it needs horticulturalist 2.0+.
 - Extract Ddocs, warm the views and deploy those staged warmed ddocs. With tests.
 - Used the `debug` framework experimentally. This is going to be revised more, probably by using console.log and console.error for high level logs / error, and then two debug levels, stage and debug, which shows timings between larger stages (ie "extracting ddocs") and individual steps

Next up:
 - Adding compaction / cleanup in the correct locations
 - Adding build stage writes to the horti deploy document
 - Changing the UI to follow and display those changes as they occur

### CODE REVIEW

**Comment by SCdF** [comment]
Ugh, looks like git didn't consider changing `legacy` -> `deploySteps` as a move, but  a copy.

Main changes in that file are the following functions: `loadStagedDdocs`, `deploySecondaryDdocs`, `deployPrimaryDdoc`, `deployStagedDdocs` (ie the ones I wrote tests for)

File: src/dbs.js [hunk 1]
```diff
@@ -5,16 +5,21 @@ PouchDB.plugin(require('pouchdb-mapreduce'));
...
+      query: unmocked,
+      remove: unmocked
     },
     builds: {
-      get: () => undefined,
-      put: () => undefined,
-      query: () => undefined
+      get: unmocked,
+      put: unmocked,
+      query: unmocked
```
Comment (garethbowen) [inline src/dbs.js]: I think we should be using sandbox.stub here so we don't have to list out each individual property. Here's my PR showing it working in api for your consideration: https://github.com/medic/medic-webapp/pull/4416/files

File: src/install/deploySteps.js [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+    info(`Deploying staged ddocs`);
+
+    return moduleWithContext._loadStagedDdocs()
+      .then(({primaryDdoc, secondaryDdocs}) => {
+        return moduleWithContext._deploySecondaryDdocs(secondaryDdocs)
+          .then(() => moduleWithContext._deployPrimaryDdoc(primaryDdoc));
+      });
+  };
+
+const updateSymlinkAndRemoveOldVersion = changedApps =>
```
Comment (garethbowen) [inline src/install/deploySteps.js]: I really don't like using the no brace style for > 1 line functions. It's really unclear (to me) that anything is being returned. Do you feel strongly about this or can we use braces for functions longer than 1 line?

File: src/install/index.js [hunk 3]
```diff
@@ -1,28 +1,27 @@
...
-// TODO: const stage = (num, message) => Promise.resolve();
-// or (num, message, fn, ...args) => Promise.resolve();
-// or (message, fn, ...args) => Promise.resolve(); (auto-num)
-
 const downloadBuild = deployDoc => {
-  info('Stage: downloading and staging install');
-  debug(`Downloading stage, getting ${keyFromDeployDoc(deployDoc)}`);
-  return DB.builds.get(keyFromDeployDoc(deployDoc), { attachments: true })
+  stage('Downloading and staging install');
+  debug(`Downloading ${keyFromDeployDoc(deployDoc)}, this may take some time...`);
```
Comment (garethbowen) [inline src/install/index.js]: What? No ellipsis?!

File: src/install/index.js [hunk 4]
```diff
@@ -40,32 +39,83 @@ const downloadBuild = deployDoc => {
...
+    `${ddoc._id.replace('_design/', '')}/${Object.keys(ddoc.views).find(k => k !== 'lib')}`;
+
+  return utils.getStagedDdocs(true)
+    .then(ddocs => {
+      debug(`Got ${ddocs.length} staged ddocs`);
+      const queries = ddocs
+        .filter(ddoc => ddoc.views && Object.keys(ddoc.views).length)
+        .map(firstView);
+
+      return probeViews(queries);
```
Comment (garethbowen) [inline src/install/index.js]: I'm so excited about this!

File: src/install/deploySteps.js (hunk 2 above)
Comment (SCdF) [reply src/install/deploySteps.js]: I think it looks worse, but I don't _strongly_ think it looks worse. Especially stuff like:

I'll try to break my programming and add braces

Comment (SCdF) [reply src/install/deploySteps.js]: I think it looks worse, but I don't _strongly_ think it looks worse. Especially stuff like:

```js
const foo = () => {
  return doStuff
    .then(doMoreStuff)
    .then(evenMoooooore)
    .then(results => {
    }
};
```

Compared to:

```js
const foo = () =>

I'll try to break my programming and add braces

**Comment by SCdF** [comment]
I'm going to merge this and I've made the requested changes in some continuing work I'm doing.

**Pull request merged**